### PR TITLE
Add XRegExp.tag in `build` plugin for tagged template string construction

### PR DESCRIPTION
**Pull request by josephfrazier**
Fixes https://github.com/slevithan/xregexp/issues/103

### CODE REVIEW

File: src/addons/build.js [hunk 1]
```diff
@@ -62,6 +62,30 @@ module.exports = function(XRegExp) {
     }
 
     /**
+     * Provides a tag function for building regexes using template literals [1]. See GitHub issue
+     * 103 for discussion [2].
+     *
+     * [1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#Tagged_template_literals
+     * [2]: https://github.com/slevithan/xregexp/issues/103
+     */
```
Comment (slevithan) [inline src/addons/build.js]: Please add an `@example` of using this to the comment

File: src/addons/build.js [hunk 2]
```diff
@@ -62,6 +62,30 @@ module.exports = function(XRegExp) {
...
+    XRegExp.tag = function (flags) {
```
Comment (slevithan) [inline src/addons/build.js]: nit: remove space after `function` to match style used by the rest of the library

File: src/addons/build.js [hunk 3]
```diff
@@ -62,6 +62,30 @@ module.exports = function(XRegExp) {
...
+        return function tag (literals /*, ...substitutions */) {
```
Comment (slevithan) [inline src/addons/build.js]: nit: remove space after `tag`. probably just remove the function name `tag` as well unless you feel this improves it.

File: src/addons/build.js [hunk 4]
```diff
@@ -62,6 +62,30 @@ module.exports = function(XRegExp) {
...
+            var substitutions = [].slice.call(arguments, 1);
```
Comment (slevithan) [inline src/addons/build.js]: Will you be converting this to ES2015 with `...substitutions` and getting rid of this?

File: src/addons/build.js [hunk 5]
```diff
@@ -62,6 +62,30 @@ module.exports = function(XRegExp) {
...
+            var subpatterns = substitutions.concat('').map(interpolate);
+            var pattern = literals.raw.map(embedSubpatternAfter).join('');
+            return XRegExp.build(pattern, subpatterns, flags);
+        };
+
+        function interpolate (substitution) {
```
Comment (slevithan) [inline src/addons/build.js]: Please move `interpolate` and `embedSubpatternAfter` out of and above `XRegExp.tag`, to the same level as `asXRegExp`. And remove the spaces after the function names. :)

File: src/addons/build.js [hunk 6]
```diff
@@ -62,6 +62,30 @@ module.exports = function(XRegExp) {
...
+     * Provides a tag function for building regexes using template literals [1]. See GitHub issue
+     * 103 for discussion [2].
+     *
+     * [1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#Tagged_template_literals
+     * [2]: https://github.com/slevithan/xregexp/issues/103
+     */
+    XRegExp.tag = function (flags) {
+        return function tag (literals /*, ...substitutions */) {
+            var substitutions = [].slice.call(arguments, 1);
+            var subpatterns = substitutions.concat('').map(interpolate);
```
Comment (slevithan) [inline src/addons/build.js]: Feels a little inelegant to always insert an additional unnecessary empty string substitution at the end of `pattern` (enabled here by `concat`-ing an empty string). I'm okay with this as is but if there's a simple way to clean this up, that might be better.

File: tests/spec/s-addons-build.js [hunk 7]
```diff
@@ -1,5 +1,81 @@
 describe('XRegExp.build addon:', function() {
 
+    describe('XRegExp.tag()', function() {
```
Comment (slevithan) [inline tests/spec/s-addons-build.js]: These tests are great--thanks!

File: src/addons/build.js (hunk 4 above)
Comment (josephfrazier) [inline src/addons/build.js]: Yes, I just wanted to make it clear for now. I'll definitely follow up with a PR that uses rest parameters across the codebase.

**Comment by josephfrazier** [comment]
Thanks for the review! I just pushed up a batch of commits that should address your comments. The one about adding an example actually prompted me to add another test and discover/fix a bug. Let me know what you think!

File: tests/spec/s-addons-build.js [hunk 8]
```diff
@@ -74,6 +74,16 @@ describe('XRegExp.build addon:', function() {
...
         });
 
+        it('should work as described in the comment @example', function() {
+            var h12 = /1[0-2]|0?[1-9]/;
+            var h24 = /2[0-3]|[01][0-9]/;
+            var hours = XRegExp.tag('x')`${h12} : | ${h24}`
+            var minutes = /^[0-5][0-9]$/;
+            var time = XRegExp.tag('x')`^ ${hours} (${minutes}) $`
+            expect(time.test('10:59')).toBe(true);
+            expect(XRegExp.exec('10:59', time).minutes).toEqual('59');
```
Comment (slevithan) [inline tests/spec/s-addons-build.js]: Nit: I'd prefer to use `toBe` wherever strict equality is appropriate, even if just for consistency (`toBe` is used one line up).

File: src/addons/build.js [hunk 9]
```diff
@@ -83,12 +88,12 @@ module.exports = function(XRegExp) {
      * var minutes = /^[0-5][0-9]$/;
      * var time = XRegExp.tag('x')`^ ${hours} (${minutes}) $`
      * time.test('10:59'); // -> true
-     * XRegExp.exec('10:59', time).minutes; // -> '59'
+     * XRegExp.exec('10:59', time)[1]; // -> '59'
```
Comment (slevithan) [inline src/addons/build.js]: Maybe just remove this line and the capturing group around `minutes` since this is no longer showing an additional feature.

**Pull request merged**